joomla/joomla-cms
# [4.3] Fix submit button focus on frontend MFA captive login page when using Webauthn

Pull Request for https://github.com/joomla/joomla-cms/pull/40330#issuecomment-1530282790 .

### Summary of Changes

When working on my pull request (PR) #40330 I've noticed that when using MFA at the site (frontend) login and the currently active MFA method is Webauthn, the javascript "com_users/js/two-factor-focus" is not loaded, and so the submit button is not activated, but when doing the same at the admin (backend) login, the script is loaded so the button is activated.

The reason for this is the `if` condition for loading the javascript in the site captive login here: https://github.com/richard67/joomla-cms/blob/4.3-dev/components/com_users/tmpl/captive/default.php#L25-L28

We do not have this condition for the admin captive login page: https://github.com/richard67/joomla-cms/blob/4.3-dev/administrator/components/com_users/tmpl/captive/default.php#L25-L26

This PR here removes this condition in file `components/com_users/tmpl/captive/default.php` and does a small code style fix in addition in that file.

In addition it fixes the `two-factor-focus.es6.js` so that it works with the above change. This basically means a revert of my PR #40330 , which has fixed the issue handled by that PR but broke the thing handled with this PR here, and replaces that wrong fix by the right one, using optional chaining as recommended by @nikosdion during review of this PR here, thanks for that.

### Testing Instructions

First test without this PR applied to reproduce the issue.

After applying the changes from this PR, run `npm ci` or `npm run build:js` when on a development environment.

When on a regular installation where you have no npm, use the update package or the custom update URL created by drone for this PR.

After having applied the changes, clear your browser cache to make sure you are not testing the old, unmodified javascript.

1. Create a user account with MFA active who can login to backend and frontend.
The user should have activated at least one MFA method which requires to enter a numeric verification code, e.g. TOTP, and another one which doesn't require such a code (Webauthn e.g with Windows Hello).
2. Login to the frontend with this account with username and password.
3. Check if depending on the MFA method, the input field for the verification code or the button to validate with the authenticator has focus.
Use the "Select a different method" button in the toolbar and check that for each method.
Watch the browser console with the developer tools of your browser to see if there are javascript errors.
Result: See section "Actual result BEFORE applying this Pull Request" without the PR, and "Expected result AFTER applying this Pull Request" with the PR.
4. Enter a verification code if necessary and use the validation button to finally log in.
5. Login to the backend with this account with username and password.
6. Check if depending on the MFA method, the input field for the verification code or the button to validate with the authenticator has focus.
Use the "Select a different method" button in the toolbar and check that for each method.
Watch the browser console with the developer tools of your browser to see if there are javascript errors.
Result: The MFA captive login page for the backend works with this PR on the same way as without this PR, i.e. nothing is broken by this PR, no javascript errors related to the "two-factor-focus" script in both cases.
7. Enter a verification code if necessary and use the validation button to finally log in.

### Actual result BEFORE applying this Pull Request

On the MFA captive login page for the frontend:
- When an MFA method is used which does not require to enter a numeric verification code, i.e. webauthn, no button has focus.
- When an MFA method is used which requires to enter a numeric verification code, e.g. TOTP or backup codes, the field to enter the verification code has focus, i.e. this works as expected.

On the MFA captive login page for the backend everything works as expected:
- When an MFA method is used which does not require to enter a numeric verification code, i.e. webauthn, the button to validate with the authenticator has focus.
- When an MFA method is used which requires to enter a numeric verification code, e.g. TOTP or backup codes, the field to enter the verification code has focus.

There are no javascript errors in the browser console related to the "two-factor-focus" script.

### Expected result AFTER applying this Pull Request

On the MFA captive login page for the frontend:
- When an MFA method is used which does not require to enter a numeric verification code, i.e. webauthn, the Submit button has focus.
- When an MFA method is used which requires to enter a numeric verification code, e.g. TOTP or backup codes, everything works as before applying this pull request, i.e. the field to enter the verification code has focus.

On the MFA captive login page for the backend: Everything works as well as before applying this pull request.

There are no javascript errors in the browser console related to the "two-factor-focus" script.

### Link to documentations
Please select:
